nexus/deployment/src/planner.rs

```diff
@@ -168,6 +162,126 @@ impl<'a> Planner<'a> {
+            // we're not allowed to use that's already running a Nexus (seems
+            // fishy!), we counted its Nexus above but will ignore it here.
+            if !sleds_ineligible_for_services.contains(&sled_id) {
+                sleds_by_num_nexus.entry(num_nexus).or_default().push(sled_id);
+            }
+        }
+
+        // TODO-correctness What should we do if we have _too many_ Nexus
+        // instances? For now, just log it the number of zones any time we have
+        // at least the minimum number.
+        let nexus_to_add =
+            self.policy.target_nexus_zone_count.saturating_sub(num_total_nexus);
+        if nexus_to_add == 0 {
+            info!(
+                self.log, "sufficient Nexus zones exist in plan";
+                "desired_count" => self.policy.target_nexus_zone_count,
+                "current_count" => num_total_nexus,
+            );
+            return Ok(());
+        }
+
+        // Ensure we have at least one sled on which we can add Nexus zones. If
+        // we don't, we have nothing else to do. This isn't a hard error,
+        // because we might be waiting for NTP on all eligible sleds (although
+        // it would be weird, since we're presumably running from within Nexus
+        // on some sled).
+        if sleds_by_num_nexus.is_empty() {
+            warn!(self.log, "want to add Nexus zones, but no eligible sleds");
+            return Ok(());
+        }
+
+        // Build a map of sled -> new nexus zone count.
+        let mut sleds_to_change: BTreeMap<Uuid, usize> = BTreeMap::new();
+
+        'outer: for _ in 0..nexus_to_add {
+            // `sleds_by_num_nexus` is sorted by key already, and we want to
+            // pick from the lowest-numbered bin. We can just loop over its
+            // keys, expecting to stop on the first iteration, with the only
+            // exception being when we've removed all the sleds from a bin.
+            for (&num_nexus, sleds) in sleds_by_num_nexus.iter_mut() {
+                // TODO choose more smartly than effectively ordering by
+                // sled_id? See
+                // https://github.com/oxidecomputer/omicron/pull/4959#discussion_r1476795735.
+                let Some(sled_id) = sleds.pop() else {
+                    // We already drained this bin; move on.
+                    continue;
+                };
+
+                // This insert might overwrite an old value for this sled (e.g.,
+                // in the "we have 1 sled and need to add many Nexus instances
+                // to it" case). That's fine.
+                sleds_to_change.insert(sled_id, num_nexus + 1);
+
+                // Put this sled back in our map, but now with one more Nexus.
+                sleds_by_num_nexus
+                    .entry(num_nexus + 1)
+                    .or_default()
+                    .push(sled_id);
+
+                continue 'outer;
```

Comment:
I'm not sure if it's any better, but it seems like you could avoid the loop label with just a `break` here and removing the `unreachable!` below.

Reply:
That feels slightly worse to me because it turns what really should be `unreachable!` (i.e., iterating all the way through this inner loop without finding a sled) into a silent fall through, which I think would result in an infinite loop.